### ISSUE TITLE
RFC: Tooltip framework and API

### ISSUE THREAD

**Issue by mcous** [Opentrons/opentrons#921]
## overview

#709 (and other tickets that involve tooltips) are blocked by the fact that we don't have tooltip components / framework in place. This RFC is a proposal for those things.

@umbhau, @IanLondon, and @Kadee80 please give this a read through and let me know how it sounds.

## proposal

This proposal comes in two parts:

- Tooltip positioning framework
- Tooltip component API

### positioning framework

The positioning framework doesn't involve the look of the tooltip, just where it is located in relation to its target. I think our best bet here is to use [Popper.js](https://popper.js.org/) - "A kickass library used to manage poppers in web applications". It seems to have everything we need:

- Positioning: `left`, `right`, `top`, `bottom`
    - Also: the ability to specify multiple locations and let it figure out which is best
- **Flipping and shifting to keep poppers in view**
- [React library available](https://github.com/souporserious/react-popper)

### tooltip component api

I can see two options for actually using tooltips in code:

1. An explicit component you use every time
2. A 🎉 Higher-order component 🎉

(1) is likely necessary no matter what, and (2) could be built on top of (1) to make our lives easier if we were so inclined.

#### Tooltip component

This probably wouldn't be so different from the tooltips we've implemented in the past. It would probably look something like:

``` js
<Tooltip
    placement='left'
    contents='Some message but could also be a React.Node or maybe null to hide the tooltip'
>
  <SomeComponent {...props} />
</Tooltip>
```

- Pro: pretty easy to know what's going on
- Con: tooltip logic/setup mixed with component logic
- Neutral: Tooltips are located in components rather than separately managed (e.g. in a central location)
- Unknown: hiding and showing logic
    - Possible: also expose a prop to manually hide and show
    - Possible: attach to DOM mouse and touch events
        - Instead of CSS
    - Possible: use CSS:hover/focus
        - Instead of DOM events
        - How would the show delay work?

#### Tooltip HOC

This idea is a little less fully baked, and would likely be done in a separate PR than the Tooltip component itself. But to address the potential downsides to using the component by itself, we could build a `withTooltip` HOC with usage something like:

``` js
type SomeComponentProps = React.ElementProps<typeof SomeComponent>
type TooltipProps = React.ElementProps<typeof Tooltip>

export default withTooltip(mapPropsToTooltip)(SomeComponent)

function mapPropsToTooltip (props: SomeComponentProps): TooltipProps {
  const tooltipProps = {}
  // ...calculate props to pass to Tooltip somehow based on component props
  // ...maybe return null if we don't want a popover
  return tooltipProps
}
```

- Pro: separates out the tooltip logic from the component's render logic
- **Pro: makes it easier to modify or even bail completely on our Tooltip component**
    - e.g. what if it turns out we hate `react-popper`?
    - e.g. what if it turns out we really do want our tooltips centrally manged?
- Pro: combined with `connect`, could make redux-driven demo-mode popovers easy
- Con: do we really need an HOC?

## estimated scope of work

Rough guesses on ticket sizes (to be more fully fleshed out if this RFC is accepted):

Implementing (1) the Tooltip component is probably a `medium` ticket, mostly because of research and reading Popper.js documentation.

Implementing (2) the `withTooltip` HOC is probably a `small` ticket, with the potential for another `small` ticket's worth of refactor work depending on how much time passes between (1) and (2) (i.e. how many places we've used `Tooltip` because `withTooltip` wasn't available).

**Comment by Kadee80** [comment]
👍 for popper.js, not that i'm pro bootstrap but bootstrap 4 uses it so its probably good for what we need and not going anywhere.
As I read the HOC part of this, I was definitely totally for it, then maybe skeptical flip flopping while reading the pros and cons. So lets maybe give it a whirl and then bail if neccessary.
